### ISSUE TITLE
IPv6 is not practically usable for production machines

### ISSUE THREAD

**Issue by JonasT** [docker/docker#13481]
As far as I'm getting from all the available documentation on docker with IPv6, it doesn't seem to be designed in a way to be usable in practice.

1.) enabling it requires specific docker daemon startup options, but does anybody even start the docker daemon manually on a production machine? Usually, the init system handles this and when I start to use things it's already running. Am I supposed to restart the whole docker daemon just to enable IPv6? Apparently, there is no /etc/docker.conf or something similar either which would be an obvious point to enable it forever.

2.) as far as I've read the IPs are randomly assigned, but there is no NAT functionality available and these are actually public IPs. (side note: to my knowledge, NAT is still possible with IPv6 albeit unusual - so this could be possible to implement at least as an option, in the future) Apparently, I can't explicitly specify a static IPv6 for a container either even if I wanted to. This all means that I see no easy way to tell how a docker container will be reachable from the outside, and I can't reliably assign a public domain to it. How can I run a public server like that?

Since we're in 2015 now and last reserves of new IPv4 addresses will run out in a few months and people are actually already causing routing conflicts over spare IPv4 addresses (see recent news articles), it would be nice if IPv6 could be fixed to be more usable for *practical* use.

**Comment by JonasT** [comment]
Since I should deploy my server in a few days, it would be nice if someone knows a workaround that is a bit simpler setting up the whole bridged network interface/all the routing stuff manually..  (that seems to be the common suggested "solution" on various websites)

**Comment by thaJeztah** [comment]
@JonasT perhaps you could discuss your use-case in the #docker-network IRC channel with the libnetwork maintainers. We're currently in feature-freeze for the 1.7 release, but discussing it could help the maintainers get a better understanding in what you think is needed.

**Comment by JonasT** [comment]
It's pretty simple really: if you assign public ips to the containers, I *NEED* a way to have a container have a predictable ip (predictable from the image, not from container creation) - e.g. by allowing me to set it myself through docker run/docker-compose.yml.

If you really want to keep random IPs, there needs to be a proper NAT in front. No NAT and random public ips is just a weird idea which isn't of any practical use - and that's the ONLY way IPv6 with docker currently operates.

And IPv6 should be possible to activate docker-wide without restarting the whole docker daemon and possibly bringing down ipv4 connections in any way.

**Comment by JonasT** [comment]
Also, all solutions to this currently seem to involve complex custom scripts and rolling my completely manually set up network bridge for my docker containers. I could do that of course - but it's just a waste of time, since docker could provide a useful implementation to start with.

**Comment by thaJeztah** [comment]
If this is for assigning an IP-address, isn't that the same as https://github.com/docker/libnetwork/issues/161? I see you commented on that already

**Comment by JonasT** [comment]
Yup it is!

**Comment by thaJeztah** [comment]
@mrjana @mavenugo any additional thoughts on this? Just checking if improvements are on the roadmap here.

**Comment by JonasT** [comment]
Guys, does no NAT also mean that all ports opened on any container with ipv6 enabled will, without any use of -p, be exposed to the internet?

Or did you at least include some basic firewall to prevent that so that port publishing still works as intended?

**Comment by jansauer** [comment]
@JonasT I like your last point. So far it seems like the plan is to just give every container a globally routable IPv6 addresses on the docker0 interface. This may lead to security problems because image authors never consider ports being accessible when not explicitly published.

At the moment the simple task of running two web servers on the same port with different ip addresses is far too difficult.

**Comment by jansauer** [comment]
@JonasT I just realised that you can predict the ipv6 address of a container by using `--mac-address`.
The ipv6 address is only a combination of the prefix set by `--fixed-cidr-v6` and the mac address set with `--mac-address`.

```` 
--fixed-cidr-v6="2001:db8:1::/64" (on the daemon)
--mac-address="02:42:ac:11:ff:ff" (on the run command)
result in: 2001:db8:1:0:0:242:ac11:ffff
```

Regarding your point about docker startup options. This is more a job for your system manufacturer than for the docker project itself. [Fedora Atomic Host](https://www.projectatomic.io/) for example uses systemd with env vars and i can set options for the docker daemon in `/etc/sysconfig/docker-network`.

At the end i'm still on your side and using ipv6 with docker should be the default and not a special use case.

**Comment by DominicBoettger** [comment]
--mac-address is nearly unusable. When restarting the container docker says that the address is already assigned.

It's needed to restart the docker daemon. I think this is really a major blocker for production use!!!

**Comment by thaJeztah** [comment]
ping @mrjana @mavenugo

**Comment by mavenugo** [comment]
@DominicBoettger @thaJeztah the issue with --mac-address, --fixed-cidr-v6 and the `address already in use` error is resolved via https://github.com/docker/docker/issues/10774. Please check the latest docker master and confirm.

**Comment by darconeous** [comment]
I fail to understand why IPv6 can't work exactly the same way that it works for IPv4: You assign a private prefix (`FD00::DEAD::BEEF/64`, for example) to the docker interface and give each container an address with this prefix. You then use NAT for getting packets in-to and out-of the container.

This doesn't preclude you from giving public IPv6 addresses to your containers if that is really what you want to do, but it will *just work* for most containers. And that's the whole idea, right? It should do something reasonable by default that just works in a way that people would expect. If you want more control, you have to dive in. Forcing everyone to do all of these extra steps to get IPv6 working is just a PITA.

You could have an option (like `--supports-v4-only`) which reverts back to the current behavior for containers which only support IPv4 (so that you can still access them via IPv6 on the host).

**Comment by timcoote** [comment]
@darconeous surely firewalls would make more sense than NAT?  I'd have thought that the host would get a /64 and allocate to the containers from that network.

shouldn't the address construction based on --fixed-cidr-v6 and --mac-address be based on SLAAC, if it's to map to the mac address or use the privacy options, otherwise. In either case, 'finding' containers would depend on using dns (?)

**Comment by JonasT** [comment]
@timcoote it doesn't really matter whether it's NAT or not, but it *does* matter that docker needs to do it automatically so the -p/publish option of docker run behaves similarly, including a protection of stuff that is not published explicitly. Everything else is just asking for people to run into unexpected security pitfalls

**Comment by Frogging101** [comment]
>It's pretty simple really: if you assign public ips to the containers, I NEED a way to have a container have a predictable ip (predictable from the image, not from container creation) - e.g. by allowing me to set it myself through docker run/docker-compose.yml.

>If you really want to keep random IPs, there needs to be a proper NAT in front. No NAT and random public ips is just a weird idea which isn't of any practical use - and that's the ONLY way IPv6 with docker currently operates.

This is the point I'm interested in too. Say I want to run a web server within Docker and have it be IPv6-compatible. How am I meant to do this if the IP address is a different random value every time I start it up? I can't route that, I can't use DNS with that, there's not much at all I can do with that honestly, aside from outgoing connections.

I understand that part of the purpose of IPv6 is to make NAT unnecessary, but these are containers. Containers are meant to be isolated from each other and from the host, and are also meant to be highly versatile in terms of access control. Giving Docker containers their own, *random* public IPv6 address does make NAT unnecessary but it also makes the containers useless in an IPv6 configuration, as well as insecure because there is no firewall on these things by default. It is expected that the host acts as access control and if the ports aren't published to the host, they won't be accessible. This, as I understand it, is not the case with IPv6 in Docker.

**Comment by timcoote** [comment]
I'm not clear why SLAAC, or even dhcpv6  would not solve the problems:
- the router to the subnet on the docker host would control access to container port/IP addresses, which are predictable (in both cases, there could be a simple mapping between mac address and the host identifier of the allocated IP address. The prefix would depend on the delegated prefix to the docker host).
- the docker host would need to be responsible for updating any network prefix changes in dns, in much the same way as it would if it was on an IPv4 subnet.

Maybe I don't understanding the problem description?

NAT is not a firewall, it just makes it much harder to ensure that the designed security has been implemented and is still working.  It also subverts mechanisms such as TLS for end to end confidentiality and integrity, and significantly complicates container <-> container connectivity as it would introduce arbitrarily deep NATting between the end points.

If the containers are intended to be used both inside a firewalled zone and directly on the internet, then they will need suitable security configuration.  There is also an argument that, even within a firewalled zone, good practice would be to not trust other services/containers.

**Comment by JonasT** [comment]
I am the original reporter, and as you can see above no the point is *not* to ask for a NAT. However, not using a NAT has various consequences which docker has apparently chose to ignore at the cost of its users which *really should be fixed*.

Those are:

- don't use random public IPs. Seriously. Nobody can DNS to that. It's just useless, don't do it. Allowing us to specify a static ip is minimum, but docker really should probably be assigning a fixed IPv6 to a container once it's created or something.

- *make sure -p/publish is still safe*. And no, telling people "uuum well even within a firewalled zone, good practice would be to not trust other services/containers" is NOT helpful. There is no big fat red warning that someone decided to drop the ball on this and unexpectedly expose all ports per default, so unless you put that everywhere you ARE gonna have people who miss out on that and you don't want to have that.

Can we now actually consider fixing this?? NAT would have been one way to address some of the problems here, you don't like it, that's understandable and I agree there are other ways that are probably better. But we shouldn't forget the actual problems over it which haven't magically disappeared and really need to be addressed.

**Comment by Frogging101** [comment]
Couldn't have said it better myself. And running two mail or web server instances on one machine with different IP addresses is exactly the sort of use case that containers should excel at. And they do, except for where it's falling flat which is the networking.

And yeah, I wasn't saying we should keep using NAT. I'm saying that NAT has the side effect of solving certain problems, and if we're taking it out of the picture then we need to have proper solutions to those problems to fill the void.

**Comment by cpuguy83** [comment]
Networking has been undergoing a major overhaul.
The reality is that the network backend in docker prior to docker 1.7 was very rigid, and could only really be used one way.
In docker 1.7 the networking was ripped out and replaced with libnetwork, which has been iterated on since then.

Docker 1.9 **should** include some major changes to how you can consume networking in containers and will include for the first time the containe rnetwork model first introduced here: https://blog.docker.com/2015/04/docker-networking-takes-a-step-in-the-right-direction-2/

Regarding stable networking (ie, assigned IP stays with the container after restart): docker/libnetwork#489

For static assignment docker/libnetwork#161

All that to say, this stuff is being worked on, and the libnetwork team is doing an amazing job.

**Comment by JonasT** [comment]
Ok so what about the -p/publish thing? Is that already being handled in IPv6 despite having no NAT? Will it be handled (so that non-published ports aren't reachable from the outer world)? This shouldn't be considered some unimportant side issue but an important security consideration.

**Comment by JonasT** [comment]
Adding another ticket about how broken docker's IPv6 is: https://github.com/docker/docker/issues/17666 publish works half-baked, but doesn't even bother to forward the correct client ip..

**Comment by wido** [comment]
I agree with @JonasT here. We need a way to have a static IPv6 address for a container.

NAT is ugly and not required, just make sure that when a container is created a static IPv6 address is assigned and recorded for that container.

When the container is started it always gets the same address.

NAT also has overhead for tracking connections and thus makes it all more complex. Please, no NAT, just plain routing. That's much faster.

**Comment by cpuguy83** [comment]
Curious, why not point at a load balancer which would be relatively static vs directly to running applications?

Also yes, definitely against the nating in general, even for ipv4 workloads.

**Comment by wido** [comment]
@cpuguy83 Not all services need to run behind a loadbalancer. For example a MariaDB database server which is accessed.

We are building our services by using IPv6-only internally. We try to avoid IPv4 and Dual-Stack where possible. In that case you want IPv6 of a container to be static and reliable.

**Comment by Deshke** [comment]
so 1.9 is out and you don't want IPv6 with NAT - just don't. We only got NAT in IPv4 because we didn't have so much addresses in the first place, IPv6 solves this problem.

* Security: ip6tables // nftables should do the trick, the host system can do the filter for the subnet or single addresses. The default should be block all incoming traffic exept the ports from the app 

* Routing: If you assign a /64 to docker your Host still does the routing and can filter connections just like in IPv4 if you got a routed network. So we only need a container startup option like "--static-ipv6=something out of the /64"

**Comment by cpuguy83** [comment]
@wido I'm not sure I understand this case as this is no different as when using ipv4 and is solved by links  or the new networking/discovery functionality.

**Comment by Frogging101** [comment]
So, I'm confused. What has changed? Are we now able to assign single consistent IP addresses to individual Docker containers? I just want to see the ability to do that, in a manner that is simple and "just works". For example "docker run --ipv4 232.53.64.5 --ipv6 2600:3a03:e030:122e::2". In the end, I hope that it won't be much more complicated than that on the user side.

**Comment by JonasT** [comment]
I checked the docker 1.9.0 release notes and it wasn't really obvious what exactly changed in regards of this issue. I wasn't able to find some sort of `--ip-address` switch for `docker run` either.

If someone could give a short hint on how the situation has supposedly improved in regards to this issue here, that would be nice and helpful. I found the new `docker network` interface which seems to allow setting up a custom IP range and all, but is there also some dead-simple way of specifying a fixed IP now? Or are people who need this expected to dig into all the details of the new network interface to do this (which I suppose supports this, but not in a totally obvious way as far as I can see)?

**Comment by bprodoehl** [comment]
1.9.0 broke the deterministic IPv6 addressing if you assign a MAC address to your containers, so anybody dealing with IPv6 stuff will almost certainly want to use the 1.9.1-rc.  Details are in #17890.

**Comment by JonasT** [comment]
Ok, but what about the new network stuff? Is there any *improvement* too?

**Comment by cpuguy83** [comment]
My point was static IP's are not a solution for container-to-container comms (even across hosts).
If all this issue is about is static ip's, this is not specific to IPv6, and there is another thread with the same discussion.

**Comment by JonasT** [comment]
@cpuguy83 it is specific to IPv6 in the way that they are public and -p/--publish is broken for IPv6, so unlike IPv4, they are actually important for proper operation. If you don't want people to use this workaround, then the docker team should make more efforts to make -p/--publish for IPv6 work properly (that is, make it properly route IPv6 addresses through instead of the current hack that does a useless source IP replacement with the host IPv4).

**Comment by Frogging101** [comment]
@cpuguy83 This issue is about not just static, but *predictable* IP addresses for containers. I want to be able to tell my container "this is your address". A single address, that I can create routing rules and point DNS records at. Going back to the original claim that "IPv6 is not practically usable for production machines", this is due to the fact that the only way to use it at present involves choosing a random IP address from a big pool, which cannot be controlled or predicted. Which is not useful for running servers, since there is no reasonable way to access it from outside.

**Comment by wido** [comment]
Same goes for linking containers. They get linked using their IPv4 address, no trace or sign of IPv6.

So currently there is no way to figure out what the IPv6 address of a container is. You need to log in and run "ip a"

**Comment by 42wim** [comment]
Not to say that fixed IPv6 addresses wouldn't be handy, but when using consul + registrator it doesn't really matter. (although for now, you have to run consul from master and patch registrator to get it all working together)

A bit more of our consul setup can be found at http://blog.42.be/2015/09/how-to-create-ipv6-only-consul-cluster.html

**Comment by stresler** [comment]
Following this issue for a while now.  Just came across [this idea](https://github.com/docker/libnetwork/issues/161#issuecomment-121518863), which sits really well with me: 

"Determine an IP from a DNS entry"

It seems a bit chicken and egg but it would, imo, solve the predictability issue if you could pass a container a hostname from which is sourced it's IP. 

e.g. 
LABEL IP-FROM-DNS=docker.com
LABEL IPV6-FROM-DNS=docker.com

This accomodates IP changes without updating run commands or docker files. It also eliminates the issue of trying to do netowrking at the application level. If an IP isn't available you'd get a reasonable error (probably because you tried to schedule it on the wrong machine). 

You can also have multiple IPs in a round robin record for supporting redundant LBs and such.

**Comment by stresler** [comment]
There was significant conversation about the user of [labels being inappropriate to use here](https://github.com/docker/docker/pull/13476#issuecomment-105890093)  "Yes, but users can pass in labels in a Dockerfile. This would in effect allow a Dockerfile writer to affect the networking of containers created from it."

I think I agree with that, but the underlying mechanism of sourcing IP config from DNS still isn't a bad idea, imo.

**Comment by wido** [comment]
@42wim True, but not all setups use Consul. For example, my home router I run some containers on. I don't use consul here locally to just run 4 containers.

**Comment by Frogging101** [comment]
With #17890, I take it we can get a predictable global IPv6 address by specifying a MAC address? That'd make Docker's IPv6 at least usable for running services.

**Comment by Deshke** [comment]
that is just a very dirty workaround in #17890

**Comment by wido** [comment]
Works for me! My containers now get the same address after every start. That's a very good start.

Now it would also be great if the IPv6 address is available in some enviroment variables for linked containers.

**Comment by aboch** [comment]
@wido @Frogging101 

FYI, the ability to get a predictable IPv6 address in the form of `--fixed-cidrv6 + --mac-address` has been there at least since docker/docker 1.5.0. Since the IPv6 capability was added to docker.

And will still be there for the default `bridge` network.

Next step to come in soon is the ability for the user to choose the IPv4 and IPv6 addresses for the container in `docker run` and `docker network connect` clis.

**Comment by robbertkl** [comment]
Like @JonasT mentions in the original issue, there is no NAT for IPv6 in Docker and the IPv6 addresses are (meant to be) publicly routable. I've created [docker-ipv6nat](https://github.com/robbertkl/docker-ipv6nat) to extend Docker with this functionality by just starting a container which handles this.

It works around some issues related to Docker IPv6 by creating a situation that is consistent with how IPv4 is handled; namely with NAT. I know NAT is unusual in an IPv6 world, but it happens to fit well in the way Docker works and how a lot of images are designed.

**Comment by cedric-dufour** [comment]
I'm currently provisioning Docker (1.11.2) for deployment into our academic (100% IPv6-ed) environment and the lack of IPv6 NAT is really troublesome.
True enough, in production scenarios, one would not want IPv6 NAT; I totally agree with that.
But in development scenario, with developers each having a docker machine and many non-production (uncharted) containers to play with, IPv6 NAT and IPv4-like ports publishing are almost mandatory. We cannot imagine having the network/DNS admins messing up with the DNS and firewall each time a developer needs to instantiate a new container.
With IPv6 NAT and IPv4-like ports publishing, each developer host has his own/single fixed IPv4/IPv6 (assigned by the network guys) and the developer is free to do whatever he wants with his containers, publishing the ports he deems fit. The mapped services are thus easily accessible, both through IPv4 and IPv6, with minimum hassles for everyone.
I think this should be implemented as an additional "--nat-v6" option, which would have the Docker daemon handle IPv6 exactly the same way as IPv4. As far as my test tell, once the '--fixed-cidr-v6' specified with a non-globally-routable IPv6 range (eg. a specific /64 ULA IPv6), it would only be a matter of applying the same ip(6)tables rules.

**Comment by justincormack** [comment]
@cedric-dufour I am planning to work on ipv6 support for docker for mac and windows soon, and as you say it is pretty impractical to assign a /64 or more per user for development, so we will use a private range internally I think. But at present docker for mac has NAT support independent of Docker, so I am not sure this means I will have to add ipv6 NAT upstream as part of this change.

**Comment by timcoote** [comment]
@cedric-dufour if your assigning ip addresses to developers won't that create issues should the code get to production, when the ip addresses are not stable? Or even issues when the network undergoes prefix renumbering?  When originally designed, I think that it was envisioned that ipv6 addresses would be stable. However, that assumption doesn't always hold in the wild.  Is it overkill to use DNS or something like Handle.net to provide long living names for sharing access points for services?

**Comment by justincormack** [comment]
@timcoote addresses in development and production will always be different, no one should be making assumptions about fixed addresses in general as services go up and down, you always should use service discovery eg DNS.

**Comment by cedric-dufour** [comment]
@timcoote I would say:
1. containers should not depend on a fixed IP address. Assigning IP addresses (and all that goes along: DNS, firewall, monitoring) should not be the container developer/user job (the only thing that should interface the two "worlds" being the protocol/port number on which a containerized service is listening). This job is that of those of manage the production environment as a whole. A simple example being: you develop a service on your own premises but your customer will make it public somewhere else; you can not expect to keep the same IP address, be it IPv4 or IPv6. Being flexible as far as IP addressing is concerned, be it IPv4 or IPv6 is a key point.
2. For plenty of reasons: I would not mix production IP addresses ranges (the one on which mission critical services are provided) and "development" IP addresses ranges (where under-development/demo/unmature servives may be accessible, be it merely to the Intranet or to some restricted Internet audience). Quality assurance policies (and product distribution policies) being a possible reason; production environments being submitted to legal requirements leading to network segregation from development environments (running those to satisfy the same requirement not being cost-effective) being a potential other, etc. (personal experience: IPs being RBL-ed because of developers screwing-up their service mail setup and the RFCs. this led us to segregate the developer and production networks, such as to prevent this kind of blunder to affect mission critical services, even if only shortly)
IPv6 is full of intents that weren't even realistic at the time it was designed (some 20 years ago); take SLAAC and RA for example: they are every network administrator nightmare in terms of security; in segments with high security requirements, those two are entirely disabled (and replaced by static or "out-of-channel" IPv6 addressing and gatewaying), despite IPv6 gurus screaming against this blasphemy. IPv6 NAT is the same kind of story.
Generally speaking, the views of system administrators, responsible for services uptime and security, and developers, seeking efficiency and flexibility, are often opposite. IPv6 NAT is not exception.

**Comment by JonasT** [comment]
Perfect IP flexibility is a silly illusion though. While the IP should never be hardcoded, at some point your domain DNS needs to meet the public IP of your server and DNS isn't inherently flexible and changeable every minute unless you use DynDNS style extensions. This is why at least some sort of basic IP stability for some containers' **public IPs** is required.. (not the internal ones behind NAT of course, which is why it's a non-issue for IPv4)

**Comment by timcoote** [comment]
@justnicormack. I agree. What I meant was that the stability of the ip addresses should be consistent between development, the CD pipeline and production

It's looking to me like IPv6 rollout will release quite a few new operational models, and these need to be understood by systems designers and implementers. In particular, there will be many ip addresses per computer, and different routes to each computer, possibly with different costs; service interactions will be much more p2p and largely unobserved. So, failover/recovery needs to be much more automatic.

I'm surprised how important ULAs have become to handle situations where a set of computers is not connected to the internet (e.g. when a building is being commissioned and the plans don't want to have a dependency on the internet.)  The model of whole subnets getting renumbered is likely to become the norm, terminating existing services.

Where existing applications have a similar pattern to this future - eg IoT home control systems - the applications jump through all sorts of hoops to get around single NAT. They mostly break with multiple NAT (e.g. CGNAT). Workarounds such as STUN/ICE are just more complexity.

@cedric-dufour in my experience the separation of responsibilities between dev and ops, creates a lot of unexpected and invisible costs. They may be inevitable in some small customers, but in larger enterprises, they can lead to a waste of around 35% of the IT spend. Where I've consulted at enterprises with highly separated networks, there has always been much increased cost of management and lots of security holes that most were unaware of. For smaller customers, I'd expect the increased cost to seriously shrink the packaged software market. But, you're correct: if that's the market you're in, then it's important ;-)

@JonasT You're correct about dynamic dns, now. But it may need to change. This isn't an issue of IPv6 itself, but it is to do with increasing connectivity.

I'm not clear why lack of expose/publish is purely a function of no NAT. Just because an address is routeable, doesn't make it reachable, and it ought to be pretty straightforward to mandate and test 'default deny' for firewalls. That said, I never cease to be surprised at how many gaps there are between corporate security policy and execution as exceptions are largely invisible.

**Comment by JonasT** [comment]
@timcoote hm your point is very valid, and if docker at least provided sensible iptables firewall rules as a default that would be a huge improvement. I guess my point about working without NAT being illusionary is wrong/overstated.

However, let me point out another use case for NAT with IPv6:

Running microservices (a more and more popular pattern) splits up a lot of stuff into tiny things, and sometimes you want those exposed under the same host = same domain = same public ip. Docker's userspace proxy is short of useless since it doesn't forward the IP which is a big security problem in itself, so we're again back at NAT for unifying multiple containers under one public ip. Of course this can be often done with running protocol-native proxies (e.g. for HTTP) on a single frontend container which routes stuff and properly retains IP and all that info, but for some protocols it might not be available and it certainly makes the setup much more complicated.

---
Also I feel like we are doing here what someone at docker should have done YEARS ago. Why isn't someone sitting down and plotting all the usage scenarios and then making a design that works better? I am happy to participate in the process, but it seems all that is done for IPv6 ever is discussion like this one on the bugtracker with not much happening... as long as nobody sits down and actually makes a unified plan and some decisions, this isn't really moving anywhere.

**Comment by Frogging101** [comment]
On 16-06-05 03:47 PM, timcoote wrote:
> I'm surprised how important ULAs have become to handle situations 
> where a set of computers is not connected to the internet (e.g. when a 
> building is being commissioned and the plans don't want to have a 
> dependency on the internet.) The model of whole subnets getting 
> renumbered is likely to become the norm, terminating existing services.
I really don't see everyone switching to dynamic IP addresses just 
because there are more ways to do it now. Static addresses are still 
simpler because you don't need all the extra infrastructure to handle 
dynamically routing them. Individuals and businesses that aren't running 
huge redundant clusters of servers won't bother with the extra 
complexity. They never did before, and there's no reason for them to do 
so now. Different configurations coexist to serve different use cases.

Currently, dynamic IP addresses are most commonly used for residential 
customers to make more efficient use of the limited IPv4 address space. 
Residential IPv6 deployments actually are more often statically assigned 
because there are enough addresses to statically assign one to every 
customer.

>  IoT home control systems
I can't believe you said IoT >.<

**Comment by cedric-dufour** [comment]
Shortly put: why all the fuss about this issue ? IPv6 NAT (and publishing/security behavior) would just be a matter of applying *exactly the same ip(6)tables rules* which are applied for IPv4 (NAT-ted by default); why not just do it (and let users deal with how they choose to implement IPv6) ? Recent version of the kernel (somewhere between 3.2 and 3.16, as far as I can tell) has everything required in place (and no need for the userland proxy). If "patches are welcome" is the answer, then I'm willing to have look into it.

**Comment by justincormack** [comment]
@cedric-dufour patches are always welcome! I think part of the issue is that ipv4 standard networking practise has changed a lot, in ways that were perhaps unexpected, with overlay networks becoming normal practise for example, while cloud providers still rarely provide ipv6, or provide very crippled implementations, so few of our customers are using it at all sadly, so consensus for the best ways to do things with ipv6 is fairly minimal.

Patches for supporting ipv6 NAT that match ipv4 NAT should be fairly straightforward and uncontroversial I would think, but if you want detailed feedback before starting work, feel free to create a new issue with a design for feedback.

**Comment by JonasT** [comment]
If you make it an additional option for the docker daemon at startup, then all those who don't want NAT in their IPv6 setup can also easily avoid it. (I would for sure appreciate it myself)

@justincormack well for me the only implementation crippled in my environment is that of docker...

**Comment by justincormack** [comment]
@JonasT I think it would have to be a per network option as there might be cases where you want a mix of routed and NAT networks. Perhaps it could be set by working out what class the addresses assigned were though. I actually can't see an option for routed ipv4, so maybe the two should be brought into line symmetrically, so both have options for routed or NAT for either address family. I do think a design document might help here.

**Comment by timcoote** [comment]
@Frogging101 I can see that I'm suffering from selection bias:
- much of the ipv6 activity that I see is on the residential side (spot the difference between weekday/weekend penetration that Google measures:http://bit.ly/1VHfuXP)
- the most prominent ipv6 promotion comes from residential facing firms (fb, google, notably NOT amazon)
- a recent experience was a huge waste of engineering effort (maybe 20% of the budget) on a residential IoT system working around NAT, and making peripatetic code too expensive.
- a common pattern for residential ipv6 is to change prefixes on router reboot. National scale measurements show that router reboots by ISPs are common, e.g. for upgrading firmware to reduce variability.

**Comment by robbertkl** [comment]
@cedric-dufour As a working example, see [docker-ipv6nat](https://github.com/robbertkl/docker-ipv6nat), which does exactly what you're suggesting: apply exactly the same NAT rules to ip6tables. This works great so far; I've been running a bunch of mail and web services, which now handle IPv6 perfectly. Access logs contain the correct addresses, SPF verification works, etc. Even though my cloud provider assigns me a large IPv6 prefix for each server, I have (statically) assigned just 1 address, just like IPv4. This IPv6 address is in my DNS and is handled by multiple containers.

It would be awesome if similar functionality would be built into Docker natively. Like @justincormack says, I agree it should be an option for each user-defined network, plus perhaps some flags for the docker daemon that is applied to the default `bridge` network.

With [docker-ipv6nat](https://github.com/robbertkl/docker-ipv6nat), I manually set a ULA prefix for each network I want to use IPv6 NAT. To keep things consistent with IPv4 automatically handing out `172.xx.` subnets, is it an idea that Docker will do this for IPv6 as well, as soon as the new IPv6 NAT option is enabled and no custom prefix is supplied with `--subnet`?

**Comment by justincormack** [comment]
@robbertkl so do you think it would be sufficient to add an `--ipv6=ula` option where docker currently accepts ipv6 addresses perhaps and have it assign a `/64` from a random ULA /48 it picks at startup?

**Comment by cedric-dufour** [comment]
@robbertkl Indeed, ULAs (RFC4193) are the closest IPv6 thing to IPv4 private ranges (RFC1918). Even though they're supposed to be globally unique, chances are slim that a conflict arises (much less than picking a IPv4 private range that does conflict with an existing numbering plan... talking out of experience :-/ )

**Comment by cedric-dufour** [comment]
@justincormack Sorry, previous comment was intended to you. On afterthoughts: maybe Docker could reserve a ULA for this purpose ?

**Comment by robbertkl** [comment]
@justincormack I'm not sure I understand, do you mean assigning the `/64` to a user-defined network when it is created?

Let me clarify:

Currently, when `--fixed-cidr` is not supplied to the docker daemon, the default `bridge` network gets a `/24` out of `172.16.0.0/12` (usually `172.17.0.0/16`). Similarly, if you create a user-defined network without `--subnet`, it again gets a `/24` out of `172.16.0.0/12` automatically.

I think the same thing could be done for IPv6, even aside from NAT: when `--fixed-cidr-v6` or `--subnet` (for docker daemon and user-defined networks respectively) are not given, but they do have a `--ipv6` flag (available for `docker network create` since 1.11.0), docker could give the network a "random" ULA range, i.e. a `/48` out of `fc00::/7`. For example, the default `bridge` network could get `fd88:8888:1::/48`, the first user-defined network will get `fd88:8888:2::/48`, etcetera, where `fd88:8888:` would be Docker's "unique" prefix from where it hands out `/48`s.

**Comment by robbertkl** [comment]
On second thought, I think it would require a separate flag after all, instead of assigning a ULA prefix by default if no subnet was given. For IPv4, it's necessary to assign a RFC1918 range to each network if no subnet is given; where else would containers get their addresses from. However, for IPv6 this is not necessary, as you could be OK with just link-local addresses.

**Comment by justincormack** [comment]
@robbertkl I think it would need a ULA subnet in order to do NAT, otherwise the container would not be able to reach the outside world at all, which would be surprising for users. But another reason to do a design review before implementation in order to get these details right.

**Comment by cedric-dufour** [comment]
I agree with @justincormack I think link-local addresses should be avoided altogether as soon as routing or multiple-hosts bridging is concerned. I'm not sure about the RFC, but I have the feeling those should not be used for anything else than their role in the various IPv6 ICMP control messages (one of those being SLAAC, NDP, RA, etc.). Doing NAT with those, or trying to create a private network spanning several hosts (e.g. by bridging a VLAN interface) would be contrary to their purpose (or might not even work).

**Comment by robbertkl** [comment]
@justincormack @cedric-dufour I was referring to the case aside from NAT, where Docker assigns networks a IPv4 private range automatically, but not for IPv6 (even if `--ipv6` is given). The reason would be because containers get link-local addresses by default.

I wasn't suggesting using link-local for NAT, sorry if that wasn't clear. NAT should definitely use ULA. Preferably a ULA prefix chosen/generated by Docker automatically when IPv6 NAT is enabled, so we don't have to do this manually (like I now have to do with [docker-ipv6nat](https://github.com/robbertkl/docker-ipv6nat) using `--fixed-cidr-v6` or `--subnet`).

And yes, +1 on a design review, exactly for issues like this.

**Comment by cpuguy83** [comment]
btw, I'm trying to get rid of iptables for NAT, replacing it with ipvs, due to potential conflicts with user-defined firewall rules and some other things.
#23022 
ipvs also supports ipv6

**Comment by cedric-dufour** [comment]
I've had a look at https://github.com/robbertkl/docker-ipv6nat . It's great work! If only it could be integrated upstream (thus voiding the requirement for 'docker run --privileged').
And if a design review is to be held, please also consider having default IPv6 security (iow. default firewall rules) match default IPv4's. It is a well known "IPv6 vulnerability" that enabling IPv6 might silently open up security holes, because it fails to match the same requirements that most consider "standard" for IPv4. And this is why corporate IPv6 deployment is such a nightmare for the network/security guys: one has to start disabling IPv6 altogether (that was some years ago, when products started to have it enabled by default) and then slowly "ramp it up" again, after careful evaluation of all its aspects (Docker being one among those).

**Comment by jgunthorpe** [comment]
I think a major consideration is missing here. I have a IPv6 network and, eg my HTTP proxy has an IPv6 (only) address.

'docker build' on my host is totally broken in this environment and there seems to be no obvious path forward. Out of the box it doesn't enable IPv6, and if I use the --ipv6 options I get a useless configuration that cannot connect to the external network.

The various --ipv6 options seem to assume that an IPv6 prefix will be delegated to the machine, which is generally pretty silly and certainly never going to happen for my laptop.

Incoming connections already seem to work fine on CoreOS - they have docker-proxy enabled so it NATs v6 to v4 connections internally (horrible, but works out of the box)

Docker needs out of the box IPv6 NAT on *outgoing* by default as the main option just to support 'docker build' *at all*.
The next option should be delegating a random SLAAC address on the prefix assigned to 'eth0' with a firewall. When ipvlan support is done this should be trivial to do.
The final and by far least desired option would be to delegate an IPv6 prefix to the machine itself, which is all that is available today.

**Comment by darconeous** [comment]
Having containers use IPv6 NAT is a no brainer. *Docker doesn't use NAT for IPv4 because of a lack of IP addresses, it uses NAT because it is convenient for the user*. We want the default case to be easy to use right? No one who is suggesting IPv6 NAT is suggesting that it be the only solution, only that it at least be an available option.

Ideally, NAT should be the default for both IPv4 and IPv6 (because it "just works") and it should be easily disabled for people who want to do something more custom, but I'd settle for having it be opt-in.

@robbertkl, why not try to push your changes upstream?

**Comment by robbertkl** [comment]
@darconeous My changes are actually not changes to Docker itself, but a separate tool/container that manages ip6tables by processing Docker events. I've done it this way for 2 reasons:

1. I wanted a quick, standalone solution that would work for most people without running custom builds of the Docker daemon; it would act as a proof-of-concept for IPv6 NAT and be usable right away (by myself and perhaps others), instead of going through a potentially long review/discussion period.
2. There seemed to be some sort of general consensus *against* NAT on IPv6; since I'm not that familiar with the Docker code base, I chose not to spend a lot of time on getting this feature in, only to risk having it shot down the minute I submit a PR - or starting a never-ending discussion.

That being said, if there *is* a real demand for IPv6 NAT, and the Docker maintainers are open to the change, I'd be happy to give it a try when I have some time. The changes in Docker itself would be quite different than what's done in [docker-ipv6nat](https://github.com/robbertkl/docker-ipv6nat), but it shouldn't be much more than duplicating all iptables modifications into similar rules for ip6tables.

**Comment by darconeous** [comment]
@robbertkl said:

> That being said, if there is a real demand for IPv6 NAT, and the Docker maintainers are open to the change, I'd be happy to give it a try when I have some time. 

I think there is demand for IPv6 support which "just works" out of the box without any additional configuration, just like how docker works with IPv4. There is only one way to accomplish that goal: **NAT**.

(There is also demand for making it easier to configure IPv6 in general, which goes beyond NAT. I don't think the two are dependent on each other.)

For what it's worth, at least one Docker maintainer (@justincormack) is [open to the idea](https://github.com/docker/docker/issues/13481#issuecomment-223900840):

> Patches for supporting ipv6 NAT that match ipv4 NAT should be fairly straightforward and uncontroversial I would think

**Comment by cedric-dufour** [comment]
IPv6 NAT can be implemented in Docker (daemon) _exactly_ the same way that IPv4.

In order to comply with the standard iptables framework I must deploy on our production hosts, I had to disable Docker management of iptables:
```
docker daemon --iptables=false ...
```
and replicate the Docker part of ip(4)tables within that framework (which isn't difficult; they're are only a few rules). Going IPv6 from threre was just _copy-paste_ to ip6tables.

The only gotcha is being able to control the IP(v4/v6) addresses assigned to running containers (for the sake of the corresponding iptables rules), which can only be achieved with non-default Docker networks. In the end, it comes down to:
```
docker network create --driver=bridge --subnet <IPv4/mask> --gateway <IPv4> --ipv6 --subnet <IPv6/mask> --gateway <IPv6> --opt com.docker.network.bridge.name=docker1 docker1
docker run --net=docker1 --ip=<IPv4> --ip6=<IPv6> ...
```
and create the Docker ip(4/6)tables for that _docker1_ interface.
PS: I opted for an IPv6 subnet within the deprecated "Site-Local" fec0::/10 (RFC 1884) range for that "Docker-Local" network (bridge)

Drawback of this method: iptables rules for for the container(s) published ports must be created manually (but this is what I want given my standard iptables framework).

My Docker Registry is fully functional using this method. 

Bottom-line:
- a working - albeight manual - solution for those who need IPv6 NAT in a more "Docker standard way"
- a proof-of-concept that IPv6 NAT can be achieved _exactly_ the same way as IPv4 NAT

A design review has been mentioned. Does starting this process require any action from us users or is it purely Docker internal stuff ?

**Comment by JonasT** [comment]
@thaJeztah there seem to be multiple people interested in getting involved and collaborating and finally getting this on track. Since you appear to be knowing your way around the docker dev team, can you help out with getting the right person to look at the most recent comments to give us all some idea on whether this can finally go ahead?

I think what is needed right now is someone confirming that the recent posts are describing an acceptable approach for whoever at docker has the final say on the related networking code, and then agreeing on how this is going to go down in more detail

**Comment by thaJeztah** [comment]
@JonasT it's currently really busy related to the 1.12 release, but let me start with adding an "needs attention" label, and putting it on the 1.13 milestone. That way it *must* be discussed during the work for 1.13 😇

**Comment by narqo** [comment]
Speaking about current Docker for Mac / Windows (e.g. `-rc2-beta17`): is IPv6-only network is supposed to be acceptable (anyhow) from within the container?

**Comment by justincormack** [comment]
We would like to support ipv6 for docker for Mac and Windows, it is on the
roadmap. Because of the setup with a VM as well it will probably have to be
set up with a private ipv6 range on the VM and then ipv6 NAT on the host,
as that is how ipv4 works. So probably not quite the same as how this issue
suggests as we would not even have a single real ipv6 address on the VM, as
that would be on the OSX or Windows host.

**Comment by jgunthorpe** [comment]
@justincormack At least on Linux you can also use proxy ARP (well, proxy neighbour discovery) with IPv6 to capture a routable SLAAC address and route it to the tunnel device for the VM. The same can be done with IPv4, but it is more difficult to get an on-demand IPv4 address.

**Comment by justincormack** [comment]
That is possible too, then we would need ipv6 NAT in the VM as well.

**Comment by jgunthorpe** [comment]
No, you don't need NAT when using proxy ARP.

You still need the private subnet with the VM, but those IPs are only used for packet routing and never appear in any public destined packets (really, they are only needed because the VM will be pretending to be Ethernet, not a tunnel device). For IPv4 this can be any private range, for IPv6 you can use the standard link-local addresses.

Once the private IP is setup, then you assign the public IP in the VM's interface with a /32 or /128 subnet and set the VM's routing table to default route all packets to the host's IP in the private range with a source of the public IP. This gets traffic out of the VM using the public IP.

On the host side the public IP is captured with proxy ARP and it routes it to the host size of virtual VM eth device, directed to the VM's private IP. This is the point you'd inject iptables as a non-NAT stateful firewall. This gets traffic into the VM using the public IP.

The host simply forwards the public IP it captures with proxy ARP/ND to the tunnel device, and forwards packets from the public IP on the tunnel device back to the network. Netfilter provides the stateful firewall, but that is for security, not for correct forwarding.

This is the basic proxy ARP configuration with the VM's virtual ethernet device pair substituted for a real ethernet network.

You don't need to use proxy ARP to get a SLAAC address and then NAT that. If you want to use NAT then just directly NAT IPv6 the same way IPv4 is being NAT'd today, as others in the thread have explained. The purpose of proxy ARP would be to give the docker container an actual public IPv6 address without having to setup IPv6 special subnets (which is largely impractical). It is similar to what ipvlan does, except it is compatible with a VM environment.

**Comment by timcoote** [comment]
@jgunthorpe I'm not clear why a host running docker (your laptop in your example) wouldn't just request an IPv6 prefix and allocate to services on containers from that address space in a deterministic way, using its firewall to constrain reachability, while allowing routability.  In normal operation, I'd expect each container to have several IPv6 addresses, depending on the route to the outside world (e.g. wifi, ethernet, LTE).  Am I thinking of the wrong problem?

**Comment by bltavares** [comment]
One example is Digital Ocean, that does not allocate a prefix for each droplet and gives you just a few ips

On July 21, 2016 5:34:55 AM GMT-04:00, timcoote <notifications@github.com> wrote:
>@jgunthorpe I'm not clear why a host running docker (your laptop in
>your example) wouldn't just request an IPv6 prefix and allocate to
>services on containers from that address space in a deterministic way,
>using its firewall to constrain reachability, while allowing
>routability.  In normal operation, I'd expect each container to have
>several IPv6 addresses, depending on the route to the outside world
>(e.g. wifi, ethernet, LTE).  Am I thinking of the wrong problem?
>
>---
>You are receiving this because you are subscribed to this thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/docker/docker/issues/13481#issuecomment-234204282

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by bprodoehl** [comment]
DigitalOcean gives you 15 usable addresses in a /64. I've asked support how I can use or request more than 15, and they said they don't support that.

**Comment by jgunthorpe** [comment]
@timecoote, I don't know what environments you work with, but in most standard IPv6 deployment there is no way to 'request a prefix', AFAIK.

Typically an Ethernet segment is assigned a /64 and the hosts on that segment choose addresses from the /64 pool and cannot go beyond that. Consider something like my corp WLAN or a coffee shop serving IPv6.

You only get to request full 128 bit addresses via SLAAC using the prefix assigned to the segment. This can be used with ipvlan or proxy ARP as I described above.

Arranging things so a single host has an entire prefix delegated to it is hard. It requires cooperation from the net-admin responsible for the segment. It requires that every host on the segment have a route entry for my docker box's prefixes (for efficiency) and that the central router(s) have the route entry too. Typically in a datacenter one would run a route daemon on each VM/docker server to propagate all the IPv6 routes and provide prefix mobility. Presumably Digital Ocean is doing that to provide /64 to containers.

I'm not able to do that on a WLAN/laptop type of environment.

**Comment by wmark** [comment]
@jgunthorpe It's actually standard in the data centers I work with to use prefix delegation. That is, you have a DUID, and use the PD flag with DHCPv6. (You have a /48 which you split according to your needs. For example, a server gets a /56 to delegate /64 to clients.) YMMV.

Anyway, flags `--ipv4=` and `--ipv6=iiii/mmm` (mmm=128 for stable addresses) for command `run` would be excellent. Not only to define a default address for exposed ports, but to set the **outgoing IP address**, too.

**Comment by timcoote** [comment]
That's how it works for domestic deployments, too. The typical house should get a /56 to carve up into 256 different physical networks so that different media qualities can easily be handled (e.g. keep the video streaming away from the low power wireless networks). The mechanisms to automate this process are under the HOMENET WG of IETF.

This isn't strictly an IPv6 issue. It's to do with the increasing address density of ubiquitous computing. However, since IPv4 cannot really handle double NAT (or even single NAT for many scenarios) with adequate performance, it only becomes apparent when you start to look at IPv6 deployments. There will be issues like picking a route to another service based on pricing that will float up into the application space.

**Comment by Frogging101** [comment]
> Anyway, flags --ipv4= and --ipv6=iiii/mmm (mmm=128 for stable addresses) for command run would be excellent. Not only to define a default address for exposed ports, but to set the source IP address, too.

Isn't that pretty much already how it works? You use `docker network create` to define a network with a subnet, an then you can use --ip or --ipv6 to assign the container an address.

**Comment by jgunthorpe** [comment]
@timecoote that still conflates deployment from the ISP with deployment within the site, I can't forsee turning on prefix delegation for my corp WLAN, for instance, but I still want 'docker build' to work out of the box in that environment with IPv6 connectivity...

**Comment by timcoote** [comment]
@jgunthorpe I would expect DHCP6-PD to work on a corporate lan.  in any case a ula prefix could be handed out by the docker daemon.  One of the challenges that Homenet addresses is how to allocate prefixes to an unconnected network, so that stuff works before the internet is connected, or when the connection is lost. I think that a lesson learned was the need for ULAs.

**Comment by jgunthorpe** [comment]
@timecoote My understanding of DHCPv6 PD is you should not allow it for untrusted clients, so no, it will not be enabled on my cell phone, coffee shop, corp wifi. If docker is the only user you can bet it will not be available widely...

How does a ULA help? We need public connectivity inside the container, not a ULA.

**Comment by JonasT** [comment]
Just reminding everyone that having multiple micro services expose stuff that needs to be reachable under one host will most likely still need NAT under v6 (the userland proxy doesn't properly retain source ips, so it's not really that useful in practice), and that in many current docker setups there are "backend" style containers that aren't supposed to be reachable from the outside which also somewhat complements putting at least some of the containers behind a NAT even with v6.

A few posts above there was a bit of a consensus that NAT for v6 as an option would at least be nice. While I can't follow all the details of the recent discussion posts due to lack of expertise, I hope you'll consider going ahead with it even if not all use cases need it.

**Comment by timcoote** [comment]
@jgunthorpe on the internet, there's no such thing as a trusted client ;-) 

I'd assumed that you were developing on your laptop using vms with containers on them in a coffee shop. So your laptop is a pd client for the wifi. the dev. vms are pd clients, either for something like homenet on your laptop, or for the wifi. Docker on the vm would handle the subnet for hosted containers.  I'm not aware of any particular vulnerabilities for dhcp6-pd. Where have you seen these mentioned?

As far as I'm aware, PD is just part of DHCP6. It's not directly available for LTE implementations, I don't think, but the phone still gets a /64, so addresses can still be handed out by phones to local devices (including containers running on the phone).

the ULA thought was for if you needed to work while not connected to the internet.

**Comment by timcoote** [comment]
@JonasT NAT is not nice ;-)  Don't confuse IP addresses with hosts. A host can have many IP addresses: the mean number in a corporate datacentre is 4.2  with IPv4 (lower for windows, higher for test machines). I'd expect that to grow with IPv6 - it certainly does for end-user devices. 

This increase is a function of increased networking demand, rather than IPv6. NATting has meant that it's not very practical to divide networks up into sensible groupings, based on physical medium characteristics (e.g. low power networks for battery powered devices separate from low latency streaming networks, free wifi, paid for LTE).

I don't think that it's a good assumption that sets of services are on the same IP address, is it? You'd lose the flexibility to split them / recombine them.

Non-reachability is a function of firewalls, not of routers, and the usual default behaviour for subnets is that they are not reachable (but they are routeable).  There's a significant hidden cost in managing the security of NATted environments as much of the environment is invisible, so there's no awareness of the actual vulnerabilities that have crept in.

I'm sure that Docker will support NAT6, but I don't think it's a good idea to use it in normal circumstances.

**Comment by JonasT** [comment]
@timcoote with hosts I mean DNS hostnames. DNS can't resolve a hostname to different IPs depending on the port, so if you make up one service under one hostname with different microservices in different containers, using publish can be vital. And the current publish implementation of docker seems to depend on NAT to work correctly - the userland proxy does the work for IPv6 too, but throwing away the source ip is a big problem so it's not really a good solution.

I'm aware NAT has a lot of downsides, but as has been discussed above it's a working solution for a lot of things that are broken right now with IPv6 and docker... I just wanted to make sure to reiterate that, since all the other fancy approaches might be nice and if they ever make everything work nicely as well I'm happy to use them eventually, but in the end of the day we need something that works.

**Comment by jgunthorpe** [comment]
@timcoote RFC3633 discusses the requirement.  DHCP is often used as a point-to-point protocol so often trust is established by controlling one side of the 'cable' (eg the FTTH), for something like WLAN the recommendation is to use RADIUS/ipsec with the DHCP protocol. This is obvious otherwise an attacker can trivially exhaust the tiny PD pool creating a DOS.

@JonasT - You are  right. docker already has the bits to manually configure PD, and this thread still exists. The desire is for something automatic, universal and 'works out of the box' - which can only be done via NATv6. For deployment an operator should pick between NATv6, ipvlan and DHCPv6 PD, they all have various trade offs.

**Comment by lcolitti** [comment]
NAT is needlessly limiting when a /64 block has 10^19 addresses. Why have 100 containers, but then put them all onto the same IP address, and then go crazy mapping ports? IPv6 has the potential to make things much simpler:

- Use a different public IPv6 address for every container; you'll never run out.
- Run a separate SSH daemon in every container, if you like; yes, they can all use port 22 on their own address. Same goes for web servers or anything else.
- Use firewalling instead of NAT. Instead of publishing ports, allow IPv6 address/port pairs, or entire IPv6 addresses for hosts that have their own firewalls. The security properties are the same, but the management overhead is reduced, and protocols that don't work well behind a NAT, like SIP or some forms of webrtc, now have a much easier time.

If you look at [draft-herbert-nvo3-ila-02](https://tools.ietf.org/html/draft-herbert-nvo3-ila-02), you'll see that one IPv6 address per container is what leading operators like Facebook are doing. (In fact, they go a step further, using the bottom 64 bits of the IPv6 address as a container ID and changing the top 64 bits as the container moves around, but let's keep it simple for now.)

This does of course require the ability to assign static IPv6 addresses to containers. That's critical, I think.

As for IPv6 address availability: RFC 7934 recommends that every customer should get access to as much IPv6 address as they need. For isolation, at least a /64 per customer is recommended; the customer does whatever they want with it. The datacenter operator can implement this via DHCPv6 PD if they want, or at worst with an RA. Each container can do ND proxying to form an address out of the /64; or you can just use bridging and SLAAC. Yes, this will work on a laptop at home.

If a VM provider doesn't assign a user at least a /64, they're needlessly limiting their customers, and said customers should ask them to follow RFC 7934.

**Comment by jgunthorpe** [comment]
@lcolitti the downside with SLAAC (eg via ipvlan, bridging, etc) is that it does not setup automatically in a multi-interface situation, eg my laptop with lan and wireless. Docker has to guess which interface should be the one to do SLAAC on and then everything breaks if the network reconfigures (eg switch to wireless from wired). NATv6 does not have these downsides. My laptop is not a datacenter, docker should not treat it like one.

**Comment by JonasT** [comment]
@lcolitti Nobody here argues NATv6 is the best solution for all use cases. However, I think it has been sufficiently demonstrated there is a need for NATv6 for some people - that doesn't mean you need to use it, and certainly nobody wants NATv6 to be the only available solution or an irreversible replacement that cannot be turned off for the current behavior.

For certain environments, especially smaller servers or laptops, NATv6 would be a huge improvement over the current *sole available behavior* which is why it should be *an option*. That datacenters run better with custom iptables instead of a NAT with fully customized routing shouldn't come to anyone's surprise.

**Comment by Frogging101** [comment]
Shouldn't this be a separate bug report? The original problem was that IPv6 was useless on production servers, due to the fact that there was no supported way to get a predictable, routable, DNSable IPv6 address. That issue has since been solved with the new "docker network create" and "docker run --ip" options, since it is now possible to statically assign a container its own IP address.

IPv6 NAT is a valid, but different feature request, and I think it would be tidier to start with a fresh bug report for that one.
--
John Brooks

**Comment by JonasT** [comment]
This is a good remark, and I made this a new feature request here: #25407

@thaJeztah please feel free to apply the necessarily labels for review and such on the new ticket, so this discussion here can stay with IPv6's general usability (or alternatively this ticket can be  closed, since as @Frogging101 pointed out the original issues discussed here are actually resolved). I hope this is ok when I'm doing this proactively and nobody hits me for taking action on this, I certainly don't want to take over tickets management or anything... but the discussion seems to indeed have gotten a bit derailed (in part by myself) in this ticket here, so it seems like a good idea

**Comment by thaJeztah** [comment]
@JonasT sounds good to me; following the discussion above; this issue can be closed?

**Comment by JonasT** [comment]
@thaJeztah as far as I'm concerned, sure

**Comment by CMCDragonkai** [comment]
Is there any documentation with regards to IPv6 for Docker for Windows `vEthernet (DockerNAT)` (the native one that uses Hyper-V)? https://docs.docker.com/engine/userguide/networking/default_network/ipv6/ assumes Linux.

**Comment by valentin2105** [comment]
https://github.com/moby/moby/issues/32675

Docker is now unusable in a IPv6 only network.
